pypicloud/cache/dynamo.py

```diff
@@ -22,6 +25,78 @@
 LOG = logging.getLogger(__name__)
 
 
+def _path_depth(package):
+    """Count the slashes in path, or return infinity on no path."""
+    path = package.data.get('path')
+    if not path:
+        return float('+inf')
+    return path.count('/')
+
+
+def _decide_between_versions(contender, current):
+    """Decide between packages with the same filename on different paths."""
+
+    if contender.last_modified < current.last_modified:
+        return contender
```

Comment:
Re: same path: Derp, yeah. Your reasoning is well guided, but I actually would defend the other way around: if they have the same path, the storage always wins. In the "same path" specific case, we're not deciding between two distinct files, but rather two different versions in time of the same file. Even if we kept the old info around, it would be describing a file we can no longer serve. In this case, there's little pypicloud can do to prevent surprises, and overwriting a path should have been prevented through bucket policy. Good catch! :)

Re: decide based on cache: I like the stability of that approach a lot, but I see too many problems with it.

Considering the cache more than just a cache, since the values that it stores get decided based on both the storage and itself, is something that asks for trouble, I would say. Ostensibly, that makes updates noncommutative (since the initial state is always an empty cache), which will make this very sensitive to timing issues. Since we're working with eventually consistent things, I'm very phobic of that.

Off the top of my head, here are some bothersome complications:
 * An "absolute reset" (blowing away the dynamo tables and re-creating them) could yield different results than rebuilding it.
 * Racy scenarios yield different cache states with no obvious reason.
 * It's so much harder to reason about in general from the explosion in possibilities (since we now have to consider all possible timelines).